### PR TITLE
nixos/modules/jenkins: add `withCli` option

### PR DESCRIPTION
**Pull request by Pamplemousse**
###### Motivation for this change

Jenkins has a CLI!
https://www.jenkins.io/doc/book/managing/cli/

This might allow more options to be added to the module later on...
(for example, the credentials has an API usable via the CLI - https://github.com/jenkinsci/credentials-plugin/blob/master/docs/user.adoc)

I am not yet too familiar with modules development. Here are two points that might be problematic with this PR.

  1. The tests for `jenkins` aren't very elaborate, so I did not try at all to test the introduced option... Should I?
  2. Probably this is pretty inefficient (downloading the CLI as part of the post start script of the related `systemd` unit). I searched into the `jenkins` war file (that is used to build `pkgs.jenkins`), but could not find the `.jar` for the cli... And the documentation clearly indicates to install the CLI by downloading the `.jar` from a running instance :confused: 

Thanks!

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

### CODE REVIEW

File: nixos/modules/services/continuous-integration/jenkins/default.nix [hunk 1]
```diff
@@ -215,10 +230,23 @@ in {
...
           sleep 1
         done
-      '';
+      '' + (lib.optionalString (cfg.withCli) ''
+
+        ${pkgs.curl.bin}/bin/curl -L -s ${jenkinsUrl}/jnlpJars/jenkins-cli.jar -o ${cfg.home}/jenkins-cli.jar
+
+        cat << __EOF__ > ${cfg.home}/jenkins-cli.sh
+        #!${pkgs.bash}/bin/bash
+        ${pkgs.jdk11}/bin/java -jar ${cfg.home}/jenkins-cli.jar -s ${jenkinsUrl} \$@
```
Comment (nagy) [inline nixos/modules/services/continuous-integration/jenkins/default.nix:242]:
```suggestion
        #!/bin/sh
        exec ${pkgs.jdk11}/bin/java -jar ${cfg.home}/jenkins-cli.jar -s ${jenkinsUrl} "\$@"
```

I dont think you need `bash` for this. Also the `$@` should better be quoted. I haven't tested it.

**Comment by Pamplemousse** [comment]
Although, I am not too sure I should be creating a `jenkins-cli` wrapper script that prefills certain CLI parameters :thinking: 
Any advice / feedback on that?

**Comment by nagy** [comment]
According to https://www.jenkins.io/doc/book/managing/cli/#using-the-client you
can set the `JENKINS_URL` environment variable, which means you would not need to `-s` parameter anymore.

You could, for example place this variable into the global environment like so:
```nix
environment.variables = {
      JENKINS_URL = jenkinsUrl;
};
```

This way, other Jenkins tools (whatever those might be) also work out-of-the-box.

You can get rid of the construction of your cli wrapper by making a custom package like this:
```nix
environment.systemPackages = [
     # ...
] ++ optionals cfg.withCli [
    # construct a wrapper jenkins-cli
    (pkgs.writeShellScriptBin "jenkins-cli" ''
        exec ${pkgs.jdk11}/bin/java -jar ${cfg.home}/jenkins-cli.jar "\$@"
    '')
];
```
This has the slight advantage, that the script cannot be modified by the jenkins user (by accident).

**Comment by Pamplemousse** [comment]
Good idea to use `JENKINS_URL`.

As for the `jenkins-cli` creation using `writeShellScriptBin`, it looks more nix-y, but only in appearance, as it creates a script in the store that relies on a file `jenkins-cli.jar` that is not...

File: nixos/modules/services/continuous-integration/jenkins/default.nix [hunk 2]
```diff
@@ -215,10 +237,21 @@ in {
...
+        #!/bin/sh
+        exec ${pkgs.jdk11}/bin/java -jar ${cfg.home}/jenkins-cli.jar "\$@"
+        __EOF__
+
+        chmod +x ${cfg.home}/jenkins-cli
+        chown ${cfg.user}:${cfg.group} ${cfg.home}/jenkins-cli*
+      '');
 
       serviceConfig = {
         User = cfg.user;
```
Comment (nagy) [inline nixos/modules/services/continuous-integration/jenkins/default.nix:257]:
```suggestion
        User = cfg.user;
        Group = cfg.group;
```

**Comment by nagy** [comment]
I find it a bit unfortunate, that we have to download this jar file every time
when we start Jenkins.

When I download the `.war` file from this release page
https://github.com/jenkinsci/jenkins/releases/tag/jenkins-2.277.4 and extract
it, I see a file called `WEB-INF/lib/cli-2.277.4.jar` with 3.3M in size. With a
bit of luck this is exactly the file that we are downloading. This means we can
put it in the nix store and remove that portion of the `postStart` script.

This change would probably be better suited into the `jenkins` derivation
itself, rather then the module. Or, because the user can change `cfg.package` we
could make a derivation on the fly which extracts this cli-jar from there. I am
still not sure where it should go.

**Comment by Pamplemousse** [comment]
I made myself the same remark (point 2. of the original PR message), but I guess I did not look hard enough at the content of the `.war` file...

Just checked it, by making sure that the `.war` file version corresponded to the one of my local running instance.
And it's a match: `diff` shows my `curl`-ed and the one extracted from the release source to be the same files!

Closing in favor of #121841 .

**Pull request closed without merging**